--- buildout.cfg.orig	2021-11-04 09:09:34.000000000 +0000
+++ buildout.cfg	2021-11-04 13:47:52.997847601 +0000
@@ -71,6 +71,9 @@
 eggs =
     Plone
     Pillow
+    bika.lims
+    baobab.lims
+    graphite.theme
 
 ############################################
 # ZCML Slugs
@@ -92,6 +95,9 @@
 #
 # Provide the *paths* to the eggs you are developing here:
 develop =
+    src/bika.lims
+    src/baobab.lims
+    src/graphite.theme
 #    src/my.package
 
 ############################################
@@ -172,10 +178,15 @@
 # config files.
 #
 [versions]
-zc.buildout = 2.5.3
-setuptools = 27.3.0
+zc.buildout = 
+setuptools = 
 Pillow = 5.1.0
+cssselect2 = 0.2.2
+soupsieve = 1.9.5
+xlsxwriter = 2.0.0
+werkzeug = 1.0.1
 
 buildout.sanitycheck = 1.0.2
 collective.recipe.backup = 4.0
 plone.recipe.unifiedinstaller = 4.3.2
+Products.DCWorkflow = 2.2.4
